### PR TITLE
check sender signature in PO mint method

### PR DESCRIPTION
**Pull request by rdinicut**
also removed isSignedWithPurpose from Identity contract
FYI: The contract is getting too big and we have problems with deployment(tests fail). #112

### CODE REVIEW

File: contracts/Identity.sol [hunk 1]
```diff
@@ -72,29 +72,4 @@ contract Identity is KeyManager {
...
     return to.call.value(value)(data);
   }
-
-  /**
-   * @dev Checks the purpose of keys used for signing.
-   * @param message bytes32 message to be verified. Must be generated with abi.encodePacked(arg1, arg2, arg3)
-   * @param signature bytes Signed data
-   * @param purpose uint256 of the key
-   */
-  function isSignedWithPurpose(
```
Comment (manuelpolzhofer) [inline contracts/Identity.sol]: Why did you remove the method?

File: contracts/erc721/UserMintableERC721.sol [hunk 2]
```diff
@@ -36,6 +42,10 @@ contract UserMintableERC721 is Initializable, ERC721, ERC721Enumerable, ERC721Me
   // Mapping from token details to token ID
   mapping(uint256 => OwnedAnchor) internal _tokenDetails;
 
+  // Value of the Signature purpose for an identity
+  // solium-disable-next-line
+  uint256 constant internal SIGNATURE_PURPOSE = 0x774a43710604e3ce8db630136980a6ba5a65b5e6686ee51009ed5f3fded6ea7e;
```
Comment (manuelpolzhofer) [inline contracts/erc721/UserMintableERC721.sol]: I think we called it signing key in some other places. We should agree on a name. :)

File: contracts/erc721/UserMintableERC721.sol [hunk 3]
```diff
@@ -430,6 +464,65 @@ contract UserMintableERC721 is Initializable, ERC721, ERC721Enumerable, ERC721Me
...
+
+
+    // Check with precise proofs if the signature is part of the documentRoot
+    require(
+      MerkleProof.verifySha256(
+        proof,
+        documentRoot,
+        sha256(abi.encodePacked(property_, signature, salt))
+      ),
+      "Signature not signed by provided identity"
```
Comment (manuelpolzhofer) [inline contracts/erc721/UserMintableERC721.sol]: Why not
```suggestion
      "provided signature is not part of the document root"
```

File: contracts/erc721/UserMintableERC721.sol [hunk 4]
```diff
@@ -430,6 +464,65 @@ contract UserMintableERC721 is Initializable, ERC721, ERC721Enumerable, ERC721Me
...
+    bytes memory signature,
+    bytes32 salt,
+    bytes32[] memory proof
+  )
+  internal
+  view
+  {
+
+    // Extract the public key from the signature
+    bytes32 pbKey_ = bytes32(
```
Comment (manuelpolzhofer) [inline contracts/erc721/UserMintableERC721.sol]: I think for external readers this always confusing.

In Ethereum the address are the first 20 bytes of the public key hash  `sha256(publicKey))`

We getting the address from `recover` method  and casting it to byte32 and calling it key.

File: contracts/erc721/UserMintableERC721.sol [hunk 5]
```diff
@@ -430,6 +464,65 @@ contract UserMintableERC721 is Initializable, ERC721, ERC721Enumerable, ERC721Me
...
+   * @param signature bytes The signature
+   * used to contract the property for precise proofs
+   * @param salt bytes32 salt for leaf construction
+   * @param proof bytes32[] proofs for leaf construction
+   */
+
+  function _requireSignedByIdentity(
+    bytes32 documentRoot,
+    address identity,
+    bytes32 signingRoot,
```
Comment (manuelpolzhofer) [inline contracts/erc721/UserMintableERC721.sol]: Where is the merkle proof that the signingRoot is part of the document?

This is very important otherwise I could just add the signature of an older document version
to it.

File: contracts/erc721/UserMintableERC721.sol [hunk 6]
```diff
@@ -430,6 +464,65 @@ contract UserMintableERC721 is Initializable, ERC721, ERC721Enumerable, ERC721Me
...
+   * @param documentRoot bytes32 the anchored document root
+   * @param identity address Identity that signed the document
+   * @param signingRoot bytes32 The message that was signed
+   * @param signature bytes The signature
+   * used to contract the property for precise proofs
+   * @param salt bytes32 salt for leaf construction
+   * @param proof bytes32[] proofs for leaf construction
+   */
+
+  function _requireSignedByIdentity(
```
Comment (manuelpolzhofer) [inline contracts/erc721/UserMintableERC721.sol]: We need a merkle proof that the identityAddress is listed as a collaborator otherwise I could just add signatures which are actually no collaborator on the document.

File: contracts/Identity.sol (hunk 1 above)
Comment (rdinicut) [reply contracts/Identity.sol]: In most cases you need the ecrecover value(Ex: In Po you build the property for precise proofs with it. Also relying on a external contract to validate a signature I think it is error prone and should validate it in the contract where you need it. The identity contract just tells you if it has a key with a specific purpose and if it's revoked. Yea you might have some code duplication but that is not important IMO. 
Also the direction of the Identity standard. It's just a proxy with an execute method

File: contracts/erc721/UserMintableERC721.sol (hunk 2 above)
Comment (rdinicut) [reply contracts/erc721/UserMintableERC721.sol]: This is the purpose not the key

Comment (manuelpolzhofer) [reply contracts/erc721/UserMintableERC721.sol]: I know. 

I just thought of renaming it to.
```suggestion
  uint256 constant internal SIGNING_PURPOSE = 0x774a43710604e3ce8db630136980a6ba5a65b5e6686ee51009ed5f3fded6ea7e;
```

or renaming the other stuff :)

File: contracts/erc721/UserMintableERC721.sol (hunk 6 above)
Comment (rdinicut) [reply contracts/erc721/UserMintableERC721.sol]: This is a helper method. It is intended to me more generic. In the PO contract we send the msg.sender as a param and validate that is a sender on the document with another method. This translates to: The sender of the invoice signed the document

File: contracts/erc721/UserMintableERC721.sol (hunk 5 above)
Comment (rdinicut) [reply contracts/erc721/UserMintableERC721.sol]: we can not generate a proof for the signingRoot and it is not necessary. We need the signingRoot just for ecrecover. We prove that the signature is part of the document.

File: contracts/erc721/UserMintableERC721.sol [hunk 7]
```diff
@@ -430,6 +464,65 @@ contract UserMintableERC721 is Initializable, ERC721, ERC721Enumerable, ERC721Me
     );
   }
 
+  /**
+   * @dev Checks that provided document is signed by an identity
```
Comment (pstehlik) [inline contracts/erc721/UserMintableERC721.sol]: `is signed by the given identity`

File: test/erc721/ERC721Document.test.js [hunk 8]
```diff
@@ -388,14 +388,39 @@ contract("UserMintableERC721", function (accounts) {
                 validRootHash,
                 sender.value,
                 signatureRoot.value,
+                signatureRoot.sorted_hashes,
```
Comment (manuelpolzhofer) [inline test/erc721/ERC721Document.test.js]: why `signatureRoot` and not `signingRoot` ?

File: contracts/erc721/UserMintableERC721.sol [hunk 9]
```diff
@@ -489,6 +491,15 @@ contract UserMintableERC721 is Initializable, ERC721, ERC721Enumerable, ERC721Me
   view
   {
 
+    require(
```
Comment (manuelpolzhofer) [inline contracts/erc721/UserMintableERC721.sol]: It might be cheaper do not use the verify method here.

Because it is just one hash function call:

something like:

```javascript
      if (signingRoot  < singingRootProof[0]) {
        require(sha256(abi.encodePacked(signingRoot, singingRootProof[0]))==documentRoot)
      } else {
       require(sha256(abi.encodePacked(singingRootProof[0], signingRoot))==documentRoot)
}
```

What do you think?

Comment (manuelpolzhofer) [reply contracts/erc721/UserMintableERC721.sol]: but I am not sure with requires in if conditions :)

Comment (rdinicut) [reply contracts/erc721/UserMintableERC721.sol]: I think it complicates the code and even if it provides a gas saving the balance will be negative.

Comment (rdinicut) [reply contracts/erc721/UserMintableERC721.sol]: the current approach is more consistent

File: contracts/erc721/UserMintableERC721.sol [hunk 10]
```diff
@@ -430,6 +464,76 @@ contract UserMintableERC721 is Initializable, ERC721, ERC721Enumerable, ERC721Me
...
+        proof,
+        documentRoot,
+        sha256(abi.encodePacked(property_, signature, salt))
+      ),
+      "Provided signature is not part of the document root"
+    );
+
+    // Check if the public key has a signature purpose on the provided identity
+    require(
+      _getIdentity(identity).keyHasPurpose(pbKey_, SIGNING_PURPOSE),
```
Comment (manuelpolzhofer) [inline contracts/erc721/UserMintableERC721.sol]: `keyHasPurpose` is not checking if the key is revoked!

Comment (rdinicut) [reply contracts/erc721/UserMintableERC721.sol]: No it does not and it will not because as I know it should return true also for revoked keys. As far as I know we are interested in the fact that identity has or had a specific key.

Comment (manuelpolzhofer) [reply contracts/erc721/UserMintableERC721.sol]: @vimukthi-git did you not talk with Razvan about the key revoking?
@rdinicut it is not always the case, we identified the problem last week and started to implement a solution for it on the node.

**Comment by rdinicut** [comment]
closing in favor of https://github.com/centrifuge/centrifuge-ethereum-contracts/pull/113
